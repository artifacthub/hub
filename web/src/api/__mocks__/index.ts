export const API = {
  getPackage: jest.fn(),
  toggleStar: jest.fn(),
  getStars: jest.fn(),
  searchPackages: jest.fn(),
  getStats: jest.fn(),
  getRandomPackages: jest.fn(),
  register: jest.fn(),
  verifyEmail: jest.fn(),
  login: jest.fn(),
  logout: jest.fn(),
  getUserProfile: jest.fn(),
  getAllRepositories: jest.fn(),
  getRepositories: jest.fn(),
  addRepository: jest.fn(),
  deleteRepository: jest.fn(),
  updateRepository: jest.fn(),
  transferRepository: jest.fn(),
  claimRepositoryOwnership: jest.fn(),
  checkAvailability: jest.fn(),
  getUserOrganizations: jest.fn(),
  getOrganization: jest.fn(),
  addOrganization: jest.fn(),
  updateOrganization: jest.fn(),
  deleteOrganization: jest.fn(),
  getOrganizationMembers: jest.fn(),
  addOrganizationMember: jest.fn(),
  deleteOrganizationMember: jest.fn(),
  confirmOrganizationMembership: jest.fn(),
  getStarredByUser: jest.fn(),
  updateUserProfile: jest.fn(),
  updatePassword: jest.fn(),
  saveImage: jest.fn(),
  getPackageSubscriptions: jest.fn(),
  addSubscription: jest.fn(),
  deleteSubscription: jest.fn(),
  getUserSubscriptions: jest.fn(),
  getWebhooks: jest.fn(),
  getWebhook: jest.fn(),
  addWebhook: jest.fn(),
  deleteWebhook: jest.fn(),
  updateWebhook: jest.fn(),
  triggerWebhookTest: jest.fn(),
  getAPIKeys: jest.fn(),
  getAPIKey: jest.fn(),
  addAPIKey: jest.fn(),
  updateAPIKey: jest.fn(),
  deleteAPIKey: jest.fn(),
  getOptOutList: jest.fn(),
  addOptOut: jest.fn(),
  deleteOptOut: jest.fn(),
  getAuthorizationPolicy: jest.fn(),
  updateAuthorizationPolicy: jest.fn(),
  getUserAllowedActions: jest.fn(),
  getSnapshotSecurityReport: jest.fn(),
  getValuesSchema: jest.fn(),
  getChangelog: jest.fn(),
  triggerTestInRegoPlayground: jest.fn(),
};
