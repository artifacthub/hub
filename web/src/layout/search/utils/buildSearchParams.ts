import { Filters } from '../../../types';
import isNull from 'lodash/isNull';

const SPECIAL_KEYS = ['text', 'page'];

interface F {
  [key: string]: string[];
}

export default (query: string): Filters => {
  const p = new URLSearchParams(query);
  let f: F = {};

  p.forEach((value, key) => {
    if (!SPECIAL_KEYS.includes(key)) {
      const values = f[key] || [];
      values.push(value);
      f[key] = values;
    }
  });

  return {
    text: p.has('text') ? p.get('text')! : undefined,
    f: { ...f },
    pageNumber: p.has('page') && !isNull(p.get('page')) ? parseInt(p.get('page')!) : 1,
  };
}
