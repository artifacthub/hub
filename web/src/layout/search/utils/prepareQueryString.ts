import isUndefined from 'lodash/isUndefined';
import { Filters } from '../../../types';

export default (filters: Filters): string => {
  const p = new URLSearchParams();
  p.set('page', filters.pageNumber.toString());
  if (!isUndefined(filters.text)) {
    p.set('text', filters.text);
  }
  if (!isUndefined(filters.f)) {
    Object.keys(filters.f).forEach((filterId: string) => {
      return filters.f[filterId].forEach((id: string) => {
        p.append(filterId, id);
      });
    });
  }
  const result = p.toString();
  return result === '' ? '' : `?${result}`;
}
