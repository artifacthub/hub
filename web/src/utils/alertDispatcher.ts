import isUndefined from 'lodash/isUndefined';

import { Alert } from '../types';

const DEFUALT_TIME = 5 * 1000; //5s

export interface AlertUpdatesHandler {
  updateAlertWrapper(alert: Alert | null): void;
}

export class AlertDispatcher {
  private updatesHandler?: AlertUpdatesHandler;
  private visibleAlert: Alert | null = null;
  private timeout?: NodeJS.Timeout;

  public subscribe(updatesHandler: AlertUpdatesHandler) {
    this.updatesHandler = updatesHandler;
  }

  public postAlert(alert: Alert | null) {
    this.visibleAlert = alert;
    this.dismissAlert();
    if (this.updatesHandler) {
      this.updatesHandler.updateAlertWrapper(this.visibleAlert);
    }
  }

  private dismissAlert() {
    this.clearTimeout();
    if (this.visibleAlert && (isUndefined(this.visibleAlert.autoClose) || this.visibleAlert.autoClose)) {
      this.timeout = setTimeout(() => {
        this.postAlert(null);
      }, this.visibleAlert.dismissOn || DEFUALT_TIME);
    }
  }

  private clearTimeout() {
    if (this.timeout) {
      clearTimeout(this.timeout);
    }
  }
}

const alertDispatcher = new AlertDispatcher();
export default alertDispatcher;
