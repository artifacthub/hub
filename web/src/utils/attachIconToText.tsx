import classnames from 'classnames';
import { isNull } from 'lodash';
import React, { useEffect, useRef, useState } from 'react';

interface Props {
  lineHeight?: number;
  className?: string;
  text: string;
  icon: JSX.Element;
  isVisible: boolean;
}

const DEFAULT_LINE_HEIGHT = 21;

export default (props: Props): JSX.Element => {
  const lineHeight: number = props.lineHeight || DEFAULT_LINE_HEIGHT;
  const ref = useRef<HTMLDivElement | null>(null);
  const icon = useRef<HTMLDivElement | null>(null);
  const lastCharacter = useRef<HTMLDivElement | null>(null);
  const [reducedFontSize, setreducedFontSize] = useState<boolean>(false);

  useEffect(() => {
    const calculateFontSize = () => {
      const { offsetTop } = ref.current!;
      const lineOfLastCharacter = Math.floor((lastCharacter.current!.offsetTop - offsetTop) / lineHeight);
      const lineOfIcon = Math.floor((icon.current!.offsetTop - offsetTop) / lineHeight);
      if (lineOfIcon !== lineOfLastCharacter) {
        setreducedFontSize(true);
      }
    };

    if (!isNull(ref.current) && props.isVisible) {
      calculateFontSize();
    }
  }, [props.isVisible, lineHeight]);

  return (
    <div
      ref={ref}
      className={props.className}
      style={{
        fontSize: reducedFontSize ? '90%' : 'inherit',
        lineHeight: `${lineHeight}px`,
      }}
    >
      {props.text.split('').map((ch: string, index: number) => {
        if (index + 1 === props.text.length) {
          return (
            <div className="d-inline" key={`ch_${index}`}>
              {ch}
            </div>
          );
        } else {
          return (
            <div className="d-inline" ref={lastCharacter} key={`ch_${index}`}>
              {ch}
            </div>
          );
        }
      })}
      <div ref={icon} className={classnames('d-inline-block', { 'position-absolute': reducedFontSize })}>
        {props.icon}
      </div>
    </div>
  );
};
