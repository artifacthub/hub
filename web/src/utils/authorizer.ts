import { isUndefined } from 'lodash';

import { API } from '../api';
import { AuthorizerAction, AuthorizerInput } from '../types';

export class Authorizer {
  private allowedActions: (AuthorizerAction | string)[] = ['all'];
  private selectedOrg?: string;
  private gettingActions: boolean = false;
  private pendingActions: (() => void)[] = [];

  public init(selectedOrg?: string) {
    this.selectedOrg = selectedOrg;
    if (selectedOrg) {
      this.getAllowedActionsList();
    }
  }

  public updateCtx(selectedOrg?: string) {
    if (isUndefined(selectedOrg)) {
      this.selectedOrg = undefined;
      this.allowedActions = ['all'];
    } else if (this.selectedOrg !== selectedOrg) {
      this.selectedOrg = selectedOrg;
      this.getAllowedActionsList();
    }
  }

  public check(input: AuthorizerInput): boolean {
    if (input.organizationName && input.organizationName !== this.selectedOrg) {
      this.selectedOrg = input.organizationName;
      this.getAllowedActionsList();
      return true;
    } else {
      if (this.gettingActions && input.onCompletion && input.organizationName) {
        this.pendingActions.push(input.onCompletion);
        return true;
      } else {
        if (isUndefined(this.selectedOrg) || this.allowedActions.includes('all')) {
          return true;
        } else {
          return this.allowedActions.includes(input.action);
        }
      }
    }
  }

  public getAllowedActionsList(onCompletion?: () => void) {
    const getUserAllowedActions = async () => {
      try {
        this.gettingActions = true;
        this.allowedActions = await API.getUserAllowedActions(this.selectedOrg!);
        this.gettingActions = false;
        if (onCompletion) {
          onCompletion();
        }
        if (this.pendingActions.length > 0) {
          this.pendingActions.forEach((action: () => void) => action());
          this.pendingActions = [];
        }
      } catch (err) {
        this.allowedActions = ['all'];
        this.gettingActions = false;
      }
    };

    if (this.selectedOrg) {
      getUserAllowedActions();
    }
  }
}

const authorizer = new Authorizer();
export default authorizer;
