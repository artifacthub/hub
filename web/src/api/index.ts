import { JSONSchema } from '@apidevtools/json-schema-ref-parser';
import camelCase from 'lodash/camelCase';
import isArray from 'lodash/isArray';
import isObject from 'lodash/isObject';
import isUndefined from 'lodash/isUndefined';

import {
  APIKey,
  APIKeyCode,
  AuthorizerAction,
  ChangeLog,
  CheckAvailabilityProps,
  Error,
  ErrorKind,
  EventKind,
  LogoImage,
  OptOutItem,
  Organization,
  OrganizationPolicy,
  Package,
  PackageStars,
  Profile,
  RegoPlaygroundPolicy,
  RegoPlaygroundResult,
  Repository,
  SearchQuery,
  SearchResults,
  SecurityReport,
  Stats,
  Subscription,
  TestWebhook,
  TsQuery,
  User,
  UserFullName,
  UserLogin,
  Webhook,
} from '../types';
import { TS_QUERY } from '../utils/data';
import getHubBaseURL from '../utils/getHubBaseURL';
import renameKeysInObject from '../utils/renameKeysInObject';

interface PackageRequest {
  packageName: string;
  repositoryKind: string;
  repositoryName: string;
  version?: string;
}

interface TransferRepositoryRequest {
  repositoryName: string;
  toOrgName?: string;
  fromOrgName?: string;
}

interface Result {
  [key: string]: any;
}

interface FetchOptions {
  method: 'POST' | 'GET' | 'PUT' | 'DELETE' | 'HEAD';
  headers?: {
    [key: string]: string;
  };
  body?: any;
}

const EXCEPTIONS = ['policies', 'rules', 'policyData', 'roles', 'crds', 'crdsExamples'];

export const toCamelCase = (r: any): Result => {
  if (isArray(r)) {
    return r.map((v) => toCamelCase(v));
  } else if (isObject(r)) {
    return Object.keys(r).reduce(
      (result, key) => ({
        ...result,
        [camelCase(key)]: EXCEPTIONS.includes(key) ? (r as Result)[key] : toCamelCase((r as Result)[key]),
      }),
      {}
    );
  }
  return r;
};

const handleErrors = async (res: any) => {
  if (!res.ok) {
    let error: Error;
    switch (res.status) {
      case 401:
        error = {
          kind: ErrorKind.Unauthorized,
        };
        break;
      case 404:
        error = {
          kind: ErrorKind.NotFound,
        };
        break;
      case 403:
        error = {
          kind: ErrorKind.Forbidden,
        };
        break;
      default:
        try {
          let text = await res.json();
          error = {
            kind: ErrorKind.Other,
            message: text.message !== '' ? text.message : undefined,
          };
        } catch {
          error = {
            kind: ErrorKind.Other,
          };
        }
    }
    throw error;
  }
  return res;
};

const handleContent = async (res: any, skipCamelConversion?: boolean) => {
  switch (res.headers.get('Content-Type')) {
    case 'text/plain; charset=utf-8':
      const text = await res.text();
      return text;
    case 'application/json':
      const json = await res.json();
      return skipCamelConversion ? json : toCamelCase(json);
    default:
      return res;
  }
};

export const apiFetch = (url: string, opts?: FetchOptions, skipCamelConversion?: boolean): any => {
  const options = opts || {};
  return fetch(url, options)
    .then(handleErrors)
    .then((res) => handleContent(res, skipCamelConversion))
    .catch((error) => Promise.reject(error));
};

export const getUrlContext = (fromOrgName?: string): string => {
  let context = '/user';
  if (!isUndefined(fromOrgName)) {
    context = `/org/${fromOrgName}`;
  }
  return context;
};

const API_BASE_URL = `${getHubBaseURL()}/api/v1`;

export const API = {
  getPackage: (request: PackageRequest): Promise<Package> => {
    let url = `${API_BASE_URL}/packages/${request.repositoryKind}/${request.repositoryName}/${request.packageName}`;
    if (!isUndefined(request.version)) {
      url += `/${request.version}`;
    }
    return apiFetch(url);
  },

  toggleStar: (packageId: string): Promise<null | string> => {
    return apiFetch(`${API_BASE_URL}/packages/${packageId}/stars`, {
      method: 'PUT',
    });
  },

  getStars: (packageId: string): Promise<PackageStars> => {
    return apiFetch(`${API_BASE_URL}/packages/${packageId}/stars`);
  },

  searchPackages: (query: SearchQuery, facets: boolean = true): Promise<SearchResults> => {
    const q = new URLSearchParams();
    q.set('facets', facets ? 'true' : 'false');
    q.set('limit', query.limit.toString());
    q.set('offset', query.offset.toString());
    if (!isUndefined(query.filters)) {
      Object.keys(query.filters).forEach((filterId: string) => {
        return query.filters[filterId].forEach((id: string) => {
          q.append(filterId, id);
        });
      });
    }
    if (!isUndefined(query.tsQueryWeb)) {
      q.set('ts_query_web', query.tsQueryWeb);
    }
    if (!isUndefined(query.tsQuery)) {
      let values: string[] = [];

      query.tsQuery.forEach((value: string) => {
        if (value !== '') {
          const activeTsQuery = TS_QUERY.find((ts: TsQuery) => ts.label === value);
          if (!isUndefined(activeTsQuery)) {
            values.push(activeTsQuery.value);
          }
        }
      });

      q.set('ts_query', values.join(' | '));
    }
    if (query.deprecated) {
      q.set('deprecated', 'true');
    }
    if (!isUndefined(query.operators) && query.operators) {
      q.set('operators', 'true');
    }
    if (!isUndefined(query.verifiedPublisher) && query.verifiedPublisher) {
      q.set('verified_publisher', 'true');
    }
    if (!isUndefined(query.official) && query.official) {
      q.set('official', 'true');
    }
    return apiFetch(`${API_BASE_URL}/packages/search?${q.toString()}`);
  },

  getStats: (): Promise<Stats> => {
    return apiFetch(`${API_BASE_URL}/packages/stats`);
  },

  getRandomPackages: (): Promise<Package[]> => {
    return apiFetch(`${API_BASE_URL}/packages/random`);
  },

  register: (user: User): Promise<null | string> => {
    const newUser = renameKeysInObject(user, { firstName: 'first_name', lastName: 'last_name' });
    return apiFetch(`${API_BASE_URL}/users`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(newUser),
    });
  },

  verifyEmail: (code: string): Promise<null> => {
    return apiFetch(`${API_BASE_URL}/users/verify-email`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        code: code,
      }),
    });
  },

  login: (user: UserLogin): Promise<null | string> => {
    return apiFetch(`${API_BASE_URL}/users/login`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        email: user.email,
        password: user.password,
      }),
    });
  },

  logout: (): Promise<null | string> => {
    return apiFetch(`${API_BASE_URL}/users/logout`);
  },

  getUserProfile: (): Promise<Profile> => {
    return apiFetch(`${API_BASE_URL}/users/profile`);
  },

  getAllRepositories: (): Promise<Repository[]> => {
    return apiFetch(`${API_BASE_URL}/repositories`);
  },

  getRepositories: (fromOrgName?: string): Promise<Repository[]> => {
    return apiFetch(`${API_BASE_URL}/repositories${getUrlContext(fromOrgName)}`);
  },

  addRepository: (repository: Repository, fromOrgName?: string): Promise<null | string> => {
    const repo = renameKeysInObject(repository, {
      displayName: 'display_name',
      authUser: 'auth_user',
      authPass: 'auth_pass',
      scannerDisabled: 'scanner_disabled',
    });
    return apiFetch(`${API_BASE_URL}/repositories${getUrlContext(fromOrgName)}`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(repo),
    });
  },

  deleteRepository: (repositoryName: string, fromOrgName?: string): Promise<null | string> => {
    return apiFetch(`${API_BASE_URL}/repositories${getUrlContext(fromOrgName)}/${repositoryName}`, {
      method: 'DELETE',
    });
  },

  updateRepository: (repository: Repository, fromOrgName?: string): Promise<null | string> => {
    const repo = renameKeysInObject(repository, {
      displayName: 'display_name',
      authUser: 'auth_user',
      authPass: 'auth_pass',
      scannerDisabled: 'scanner_disabled',
    });
    return apiFetch(`${API_BASE_URL}/repositories${getUrlContext(fromOrgName)}/${repository.name}`, {
      method: 'PUT',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(repo),
    });
  },

  transferRepository: (params: TransferRepositoryRequest): Promise<null | string> => {
    return apiFetch(
      `${API_BASE_URL}/repositories${getUrlContext(params.fromOrgName)}/${params.repositoryName}/transfer${
        isUndefined(params.toOrgName) ? '' : `?org=${params.toOrgName}`
      }`,
      {
        method: 'PUT',
      }
    );
  },

  claimRepositoryOwnership: (repo: Repository, toOrgName?: string): Promise<null | string> => {
    return apiFetch(
      `${API_BASE_URL}/repositories${getUrlContext(repo.organizationName || undefined)}/${repo.name}/claimOwnership${
        isUndefined(toOrgName) ? '' : `?org=${toOrgName}`
      }`,
      {
        method: 'PUT',
      }
    );
  },

  checkAvailability: async (props: CheckAvailabilityProps): Promise<boolean> => {
    return fetch(`${API_BASE_URL}/check-availability/${props.resourceKind}?v=${encodeURIComponent(props.value)}`, {
      method: 'HEAD',
    })
      .then((res: any) => {
        switch (res.status) {
          case 404:
            return Promise.resolve(false);
          default:
            return Promise.resolve(true);
        }
      })
      .catch(() => {
        return Promise.resolve(true);
      });
  },

  getUserOrganizations: (): Promise<Organization[]> => {
    return apiFetch(`${API_BASE_URL}/orgs/user`);
  },

  getOrganization: (organizationName: string): Promise<Organization | null> => {
    return apiFetch(`${API_BASE_URL}/orgs/${organizationName}`);
  },

  addOrganization: (organization: Organization): Promise<null | string> => {
    const org = renameKeysInObject(organization, {
      displayName: 'display_name',
      logoImageId: 'logo_image_id',
      homeUrl: 'home_url',
    });
    return apiFetch(`${API_BASE_URL}/orgs`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(org),
    });
  },

  updateOrganization: (organization: Organization, name: string): Promise<null | string> => {
    const org = renameKeysInObject(organization, {
      displayName: 'display_name',
      logoImageId: 'logo_image_id',
      homeUrl: 'home_url',
    });
    return apiFetch(`${API_BASE_URL}/orgs/${name}`, {
      method: 'PUT',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(org),
    });
  },

  deleteOrganization: (orgName: string): Promise<null | string> => {
    return apiFetch(`${API_BASE_URL}/orgs/${orgName}`, {
      method: 'DELETE',
    });
  },

  getOrganizationMembers: (organizationName: string): Promise<User[]> => {
    return apiFetch(`${API_BASE_URL}/orgs/${organizationName}/members`);
  },

  addOrganizationMember: (organizationName: string, alias: string): Promise<null | string> => {
    return apiFetch(`${API_BASE_URL}/orgs/${organizationName}/member/${encodeURI(alias)}`, {
      method: 'POST',
    });
  },

  deleteOrganizationMember: (organizationName: string, alias: string): Promise<null | string> => {
    return apiFetch(`${API_BASE_URL}/orgs/${organizationName}/member/${encodeURI(alias)}`, {
      method: 'DELETE',
    });
  },

  confirmOrganizationMembership: (organizationName: string): Promise<null> => {
    return apiFetch(`${API_BASE_URL}/orgs/${organizationName}/accept-invitation`);
  },

  getStarredByUser: (): Promise<Package[]> => {
    return apiFetch(`${API_BASE_URL}/packages/starred`);
  },

  updateUserProfile: (profile: UserFullName): Promise<null | string> => {
    const updatedProfile = renameKeysInObject(profile, {
      firstName: 'first_name',
      lastName: 'last_name',
      profileImageId: 'profile_image_id',
    });
    return apiFetch(`${API_BASE_URL}/users/profile`, {
      method: 'PUT',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(updatedProfile),
    });
  },

  updatePassword: (oldPassword: string, newPassword: string): Promise<null | string> => {
    return apiFetch(`${API_BASE_URL}/users/password`, {
      method: 'PUT',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        old: oldPassword,
        new: newPassword,
      }),
    });
  },

  saveImage: (data: string | ArrayBuffer): Promise<LogoImage> => {
    return apiFetch(`${API_BASE_URL}/images`, {
      method: 'POST',
      body: data,
    });
  },

  getPackageSubscriptions: (packageId: string): Promise<Subscription[]> => {
    return apiFetch(`${API_BASE_URL}/subscriptions/${packageId}`);
  },

  addSubscription: (packageId: string, eventKind: EventKind): Promise<string | null> => {
    return apiFetch(`${API_BASE_URL}/subscriptions`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        package_id: packageId,
        event_kind: eventKind,
      }),
    });
  },

  deleteSubscription: (packageId: string, eventKind: EventKind): Promise<string | null> => {
    return apiFetch(`${API_BASE_URL}/subscriptions?package_id=${packageId}&event_kind=${eventKind}`, {
      method: 'DELETE',
    });
  },

  getUserSubscriptions: (): Promise<Package[]> => {
    return apiFetch(`${API_BASE_URL}/subscriptions`);
  },

  getWebhooks: (fromOrgName?: string): Promise<Webhook[]> => {
    return apiFetch(`${API_BASE_URL}/webhooks${getUrlContext(fromOrgName)}`);
  },

  getWebhook: (webhookId: string, fromOrgName?: string): Promise<Webhook> => {
    return apiFetch(`${API_BASE_URL}/webhooks${getUrlContext(fromOrgName)}/${webhookId}`);
  },

  addWebhook: (webhook: Webhook, fromOrgName?: string): Promise<null | string> => {
    const formattedWebhook = renameKeysInObject(webhook, { contentType: 'content_type', eventKinds: 'event_kinds' });
    const formattedPackages = webhook.packages.map((packageItem: Package) => ({
      package_id: packageItem.packageId,
    }));
    return apiFetch(`${API_BASE_URL}/webhooks${getUrlContext(fromOrgName)}`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        ...formattedWebhook,
        packages: formattedPackages,
      }),
    });
  },

  deleteWebhook: (webhookId: string, fromOrgName?: string): Promise<null | string> => {
    return apiFetch(`${API_BASE_URL}/webhooks${getUrlContext(fromOrgName)}/${webhookId}`, {
      method: 'DELETE',
    });
  },

  updateWebhook: (webhook: Webhook, fromOrgName?: string): Promise<null | string> => {
    const formattedWebhook = renameKeysInObject(webhook, { contentType: 'content_type', eventKinds: 'event_kinds' });
    const formattedPackages = webhook.packages.map((packageItem: Package) => ({
      package_id: packageItem.packageId,
    }));
    return apiFetch(`${API_BASE_URL}/webhooks${getUrlContext(fromOrgName)}/${webhook.webhookId}`, {
      method: 'PUT',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({ ...formattedWebhook, packages: formattedPackages }),
    });
  },

  triggerWebhookTest: (webhook: TestWebhook): Promise<string | null> => {
    const formattedWebhook = renameKeysInObject(webhook, { contentType: 'content_type', eventKinds: 'event_kinds' });

    return apiFetch(`${API_BASE_URL}/webhooks/test`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(formattedWebhook),
    });
  },

  getAPIKeys: (): Promise<APIKey[]> => {
    return apiFetch(`${API_BASE_URL}/api-keys`);
  },

  getAPIKey: (apiKeyId: string): Promise<APIKey> => {
    return apiFetch(`${API_BASE_URL}/api-keys/${apiKeyId}`);
  },

  addAPIKey: (name: string): Promise<APIKeyCode> => {
    return apiFetch(`${API_BASE_URL}/api-keys`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        name: name,
      }),
    });
  },

  updateAPIKey: (apiKeyId: string, name: string): Promise<string | null> => {
    return apiFetch(`${API_BASE_URL}/api-keys/${apiKeyId}`, {
      method: 'PUT',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({ name: name }),
    });
  },

  deleteAPIKey: (apiKeyId: string): Promise<string | null> => {
    return apiFetch(`${API_BASE_URL}/api-keys/${apiKeyId}`, {
      method: 'DELETE',
    });
  },

  getOptOutList: (): Promise<OptOutItem[]> => {
    return apiFetch(`${API_BASE_URL}/subscriptions/opt-out`);
  },

  addOptOut: (repositoryId: string, eventKind: EventKind): Promise<string | null> => {
    return apiFetch(`${API_BASE_URL}/subscriptions/opt-out`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        repository_id: repositoryId,
        event_kind: eventKind,
      }),
    });
  },

  deleteOptOut: (optOutId: string): Promise<string | null> => {
    return apiFetch(`${API_BASE_URL}/subscriptions/opt-out/${optOutId}`, {
      method: 'DELETE',
    });
  },

  getAuthorizationPolicy: (orgName: string): Promise<OrganizationPolicy> => {
    return apiFetch(`${API_BASE_URL}/orgs/${orgName}/authorizationPolicy`);
  },

  updateAuthorizationPolicy: (orgName: string, policy: OrganizationPolicy): Promise<string | null> => {
    const formattedPolicy = renameKeysInObject(
      { ...policy },
      {
        authorizationEnabled: 'authorization_enabled',
        predefinedPolicy: 'predefined_policy',
        customPolicy: 'custom_policy',
        policyData: 'policy_data',
      }
    );

    return apiFetch(`${API_BASE_URL}/orgs/${orgName}/authorizationPolicy`, {
      method: 'PUT',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(formattedPolicy),
    });
  },

  getUserAllowedActions: (orgName: string): Promise<AuthorizerAction[]> => {
    return apiFetch(`${API_BASE_URL}/orgs/${orgName}/userAllowedActions`);
  },

  getSnapshotSecurityReport: (packageId: string, version: string): Promise<SecurityReport> => {
    return apiFetch(`${API_BASE_URL}/packages/${packageId}/${version}/securityReport`, undefined, true);
  },

  getValuesSchema: (packageId: string, version: string): Promise<JSONSchema> => {
    return apiFetch(`${API_BASE_URL}/packages/${packageId}/${version}/valuesSchema`, undefined, true);
  },

  getChangelog: (packageId: string): Promise<ChangeLog[]> => {
    return apiFetch(`${API_BASE_URL}/packages/${packageId}/changelog`);
  },

  // External API call
  triggerTestInRegoPlayground: (data: RegoPlaygroundPolicy): Promise<RegoPlaygroundResult> => {
    return apiFetch('https://play.openpolicyagent.org/v1/share', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(data),
    });
  },
};
