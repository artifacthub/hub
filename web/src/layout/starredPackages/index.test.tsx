import { render, waitFor } from '@testing-library/react';
import React from 'react';
import { BrowserRouter as Router } from 'react-router-dom';
import { mocked } from 'ts-jest/utils';

import { API } from '../../api';
import { ErrorKind, Package } from '../../types';
import StarredPackagesView from './index';
jest.mock('../../api');

const mockHistoryPush = jest.fn();

jest.mock('react-router-dom', () => ({
  ...(jest.requireActual('react-router-dom') as {}),
  useHistory: () => ({
    push: mockHistoryPush,
  }),
}));

const getMockStarredPackages = (fixtureId: string): Package[] => {
  return require(`./__fixtures__/index/${fixtureId}.json`) as Package[];
};

describe('StarredPackagesView', () => {
  afterEach(() => {
    jest.resetAllMocks();
  });

  it('creates snapshot', async () => {
    const mockPackages = getMockStarredPackages('1');
    mocked(API).getStarredByUser.mockResolvedValue(mockPackages);

    const result = render(
      <Router>
        <StarredPackagesView />
      </Router>
    );

    await waitFor(() => {
      expect(API.getStarredByUser).toHaveBeenCalledTimes(1);
      expect(result.asFragment()).toMatchSnapshot();
    });
  });

  describe('Render', () => {
    it('renders component', async () => {
      const mockPackages = getMockStarredPackages('2');
      mocked(API).getStarredByUser.mockResolvedValue(mockPackages);

      const { getByText } = render(
        <Router>
          <StarredPackagesView />
        </Router>
      );

      await waitFor(() => {
        expect(API.getStarredByUser).toHaveBeenCalledTimes(1);
      });
      expect(getByText('Your starred packages')).toBeInTheDocument();
    });
  });

  describe('Packages', () => {
    it('renders 5 packages', async () => {
      const mockPackages = getMockStarredPackages('4');
      mocked(API).getStarredByUser.mockResolvedValue(mockPackages);

      const { getAllByRole } = render(
        <Router>
          <StarredPackagesView />
        </Router>
      );

      const packages = await waitFor(() => getAllByRole('listitem'));
      expect(packages).toHaveLength(5);
    });

    it('displays no data component when no packages', async () => {
      const mockPackages = getMockStarredPackages('5');
      mocked(API).getStarredByUser.mockResolvedValue(mockPackages);

      const { getByTestId } = render(
        <Router>
          <StarredPackagesView />
        </Router>
      );

      const noData = await waitFor(() => getByTestId('noData'));
      expect(noData).toBeInTheDocument();
      expect(noData).toHaveTextContent('You have not starred any package yet');
    });

    it('renders error message when getStarredByUser call fails with not unauthorized error', async () => {
      mocked(API).getStarredByUser.mockRejectedValue({ kind: ErrorKind.Other });

      const { getByTestId } = render(
        <Router>
          <StarredPackagesView />
        </Router>
      );

      const noData = await waitFor(() => getByTestId('noData'));
      expect(noData).toBeInTheDocument();
      expect(noData).toHaveTextContent(/An error occurred getting your starred packages, please try again later./i);
    });

    it('calls history push to load login modal when user is not signed in', async () => {
      mocked(API).getStarredByUser.mockRejectedValue({
        kind: ErrorKind.Unauthorized,
      });

      render(
        <Router>
          <StarredPackagesView />
        </Router>
      );

      await waitFor(() => {
        expect(mockHistoryPush).toHaveBeenCalledTimes(1);
        expect(mockHistoryPush).toHaveBeenCalledWith('/?modal=login&redirect=/packages/starred');
      });
    });
  });
});
