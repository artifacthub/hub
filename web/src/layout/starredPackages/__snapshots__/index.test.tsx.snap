// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StarredPackagesView creates snapshot 1`] = `
<DocumentFragment>
  <div
    class="d-flex flex-column flex-grow-1 position-relative"
  >
    <div
      class="position-absolute p-5 wrapper undefined"
    >
      <div
        class="d-flex flex-row align-items-center justify-content-center w-100 h-100"
      >
        <div
          class="spinner-border text-primary spinner undefined"
          role="status"
        >
          <span
            class="sr-only"
          >
            Loading...
          </span>
        </div>
      </div>
    </div>
    <main
      class="container-lg px-sm-4 px-lg-0 py-5"
      role="main"
    >
      <div
        class="flex-grow-1 position-relative"
      >
        <div
          class="h3 pb-0"
        >
          <div
            class="d-flex align-items-center justify-content-center"
          >
            <div>
              Your starred packages
            </div>
          </div>
        </div>
        <div
          class="row mx-auto mt-4 wrapper"
        />
      </div>
    </main>
  </div>
</DocumentFragment>
`;
